### ISSUE TITLE
App won't open

### ISSUE THREAD

**Issue by SirKittyton** [terrymacdonald/DisplayMagician#123]
Hello!

I recently discovered this app and it quickly solved my issues with displays for games. 

After a reboot of my system, the app no longer will open, despite a fresh install and reboot, it will no longer open.

I can see when I attempt to start the application, it pops up in task manager but immediately crashes.

On Windows 10.

Thanks

**Comment by terrymacdonald** [comment]
Hi @SirKittyton 

I think some other users have struck this issue. Can you please try this new test version DisplayMagician v2.4.0.26, to see if it fixes your problem?

To install DisplayMagician v2.4.0.26, please download this zip file and install the exe inside it. Please test the software and make sure that it now swaps display profiles without any errors: 
[DisplayMagicianSetup-v2.4.0.26.zip](https://github.com/terrymacdonald/DisplayMagician/files/8869771/DisplayMagicianSetup-v2.4.0.26.zip)

Thanks
Terry

**Comment by SomeKay** [comment]
Hey @terrymacdonald, here are my findings:
- yesterday, I first did an install of the new (2.4.0.26) version over the current official version, restarted and it didn't want to start up
- after your comment, I uninstalled it again, restarted and installed only the new version and restarted again - it started working then and changed the profiles between Surround and separate displays wonderfully
- today, I've booted my laptop up again and DisplayMagician won't start again and I see the same errors in Event Log

But, here's something that did help and fixed it:
- when I boot up my laptop, I am back to 3 displays (not in Surround) and the App won't start
- when I use Nvidia's Control Panel and enable Surround again, the App starts up successfully!
- I can change my profiles normally now

So, in summary: Surround has to be enabled and running for DisplayMagician to start. Otherwise, I get the Event Log errors I pasted in my previous comment.

**Comment by terrymacdonald** [comment]
Hi,

You have to delete your old display profiles, and then create them again.
The reason is that the display profile file format changed, and the old
display profiles don't have all the information in them that the new
version of DisplayMagician needs.

I strongly recommend deleting your old profiles, then recreate then again.
Once you've done that, you will need to update each game shortcut to use
the new display profiles you created.

Please note that this is a test version of DisplayMagician, and therefore
you will have to delete and recreate your display profiles again when the
actual final 2.4.0 version of DisplayMagician is released. This is because
there is another little change to the file format required to fix another
little big we've recently found.

Anyway, that should be enough information to get you going again without
any troubles.

Thanks
Terry

**Comment by SomeKay** [comment]
I've tried now uninstalling DisplayMagician completely, deleting all profiles, deleting every trace of it (in %APPDATA%) and then reinstalling the test 2.4.0 version.

It's still presenting the same issues: it doesn't want to start (and crashes with the same error in Event Viewer) until I go into Nvidia Control Panel and enable Surround.

There is one additional oddity, but maybe that's how it is supposed to be: when I'm creating a profile for 3 separate displays (so no Surround so I can support games which run Windowed Mode triple display like iRacing), DisplayMagician still shows it as one big display (see screenshot). Is that normal?

![image](https://user-images.githubusercontent.com/54692/174153557-c80ef1a6-5547-4d7a-8832-721c04a14ef9.png)

**Comment by terrymacdonald** [comment]
Hi,

No, that's not normal. It looks as though NVIDIA surround needs to be
turned off via NVIDIA control panel.

I would shutdown DisplayMagician, then open NVIDIA control panel and
confirm that NVIDIA surround is turned off, then save your triple config in
DisplayMagician.

One thing to note is that the windows display settings window sometimes
needs to be restarted to pick up the current display settings. This is true
if you turn on NVIDIA surround, and of you turn off NVIDIA surround.

Another thing to note is that the NVIDIA control panel surround checkbox
also won't update sometimes, and you need to close and reopen NVIDIA
control panel for it to her the correct settings. The reason for this is
that NVIDIA control panel expects the only way you can change surround
settings is via their application. But as DisplayMagician talks directly to
the API it bypasses the NVIDIA control panel, meaning that the NVIDIA
control panel doesn't always reflect the truth (but it does once you
restart NVIDIA control panel).

Thanks
Terry

**Comment by terrymacdonald** [comment]
@SomeKay I forgot to say, can you please zip up all of the 5 DisplayMagician.log files in C:\Users\<username>\AppData\Local\DisplayMagician\Logs\ so that I can see what errors are in them? It may help me fix the issue.

Thanks
Terry

**Comment by SomeKay** [comment]
@terrymacdonald Of course, the logs from the version which doesn't work (2.4.0.26) are attached.
[DisplayMagicianLogs.zip](https://github.com/terrymacdonald/DisplayMagician/files/8943467/DisplayMagicianLogs.zip)

On a side note: I tried installing an old version (2.1.5) and that one works perfectly.

**Comment by terrymacdonald** [comment]
@SomeKay some of the errors in the DisplayMagician logs **should** be fixed in the latest test version of DisplayMagician which I'll attach here. 

There are some really weird issues in the logs as, sometimes the log files are only 1 line long! I've never seen that ever before. It's really very strange.

I'm going to add some logs to the codebase and I'll make a new test version just for you now that creates more log entries near the startup phase. And here it is:
[DisplayMagicianSetup-v2.4.0.50.zip](https://github.com/terrymacdonald/DisplayMagician/files/8946550/DisplayMagicianSetup-v2.4.0.50.zip)

Please download it, unzip it and install it over the top of the previous version of DisplayMagician v2.4.0 you were using. 
IMPORTANT: Please delete your previous display profiles if DisplayMagician does finally start up. If it doesn't start up, then please reboot your computer and try again. If it still doesn't start up, then please zip up the DisplayMagician log files again, and upload here again. I'll hopefully get a better idea of where it is failing this time! 🤞 

Thanks
Terry

**Comment by terrymacdonald** [comment]
@SirKittyton Can you please try again with DisplayMagician v2.4.0.50 too? See above comment ^

Thanks
Terry

**Comment by SomeKay** [comment]
@terrymacdonald Unfortunately, the new version doesn't do anything new. I still can't start on separate triples, but can start it when on Nvidia Surround.

Here are the 2 logs named appropriately:
[DisplayMagicianLogs.zip](https://github.com/terrymacdonald/DisplayMagician/files/8952354/DisplayMagicianLogs.zip)

**Comment by terrymacdonald** [comment]
Oh I forgot! Please also set the Logging level to TRACE. You can do that in the Settings window. Most logs are set at Trace logging level, so I expect we're missing a lot of info at the moment.

If you cannot start DisplayMagician at all, then please edit the Settings_2.4.json file in c:\Users\<username>\AppData\Local\DisplayMagician folder, and change the `"LogLevel": "Info",` line to `"LogLevel": "Trace",`
This will tell DisplayMagician to log more information.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@SomeKay Once you've done the above, please try restarting DisplayMagician, and you should get some nicely detailed logs that will help me figure out what bit is broken.

Please generate the Support ZIP File if you can (if DisplayMagician starts). DisplayMagician Support ZIP Files actually contain the last 4 log files, as well as the Display Profiles, and the Game Shortcuts, so will give me exactly the information I need... once the logging level has been changed to TRACE :).

Thanks
Terry

**Comment by terrymacdonald** [comment]
@SirKittyton Did DisplayMagician 2.4.0.50 fix the issue? [DisplayMagicianSetup-v2.4.0.50.zip](https://github.com/terrymacdonald/DisplayMagician/files/8946550/DisplayMagicianSetup-v2.4.0.50.zip)

If so, please reply here so I know I can consider your issue closed. If you are still having an issue please create a Support ZIP File from the DisplayMagician settings window, and then upload it here. If you cannot start DisplayMagician at all, then please edit the Settings_2.4.json file in c:\Users<username>\AppData\Local\DisplayMagician folder, and change the "LogLevel": "Info", line to "LogLevel": "Trace", and then restart DisplayMagician. This will tell DisplayMagician to log more information which we can check to see where the error is.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@SomeKay @SirKittyton confirmed it was the Windows DPI scaling. From my testing it seems some hardware (most likely laptops) return data outside the expected range. I've now made DisplayMagician tell windows what DPI awareness it has, so that we get the guaranteed right DPI settings, and I've made WinLibrary handle expected values (as in my testing these values still result in the correct Windows DPI Scale being applied). 

I've also adjusted the process for adding a '.old' to the end of the older json config files so that it does it to all of them reliably. This should help avoid DisplayMagician wanting to patch the wrong things.

I've also added a Windows display path processing improvement that should help on devices with a lot of display paths that have accumulated over time. Display Paths are the bits in windows that link a display source (a particular desktop) with a display target (a particular monitor). Windows keeps track of every monitor it's ever seen, and that eventually leaves a lot of stale entries in the list. We now try to skip over the ones that can't be used. 

I've wrapped this all up into a new test version of DisplayMagician v2.4..0.64 that *should* fix the issue you're experiencing. To test this, please perform the following:

- Download DisplayMagician-v2.4..0.64.zip from here: [DisplayMagicianSetup-v2.4.0.64.zip](https://github.com/terrymacdonald/DisplayMagician/files/8996434/DisplayMagicianSetup-v2.4.0.64.zip)
- Unzip the zip file
- Run the DisplayMagician-v2.4..0.64.msi
- See if it loads properly!
- Go to the Display Profiles window
- Delete any old Display Profiles listed in there
- Use Windows Setup or NVIDIA Control Panel or AMD Adrenalin to set up a display layout you want
- Swap back to DisplayMagician and save the display layout as a Display Profile.
- Swap back to Windows Setup or NVIDIA Control Panel or AMD Adrenalin to set up another display layout you want
- Swap back to DisplayMagician and save the display layout as a Display Profile.

Once you have finished setting up your Display Profiles, please try to swap between them *using the Apply button the Display Profiles window*. This is critical as I want to test that everything works using the simplest way to do so.

Please then create a Support ZIP File by going click the Settings button on the Main page, and then clicking on the Create Support ZIP File button. Please upload the ZIP file using you make to this issue using your Web Browser, so that I can see if this fixed the problem for you.

Hope it works!

Thanks
Terry

**Comment by terrymacdonald** [comment]
Hi @SirKittyton @SomeKay,

I can't wait any longer, so I'm releasing DisplayMagician v2.4.0 today which *should* fix you issue. I'm going to close your issue for now so I know what other issues I need to work on, but if you do some testing and this same issue is still causing you a problem, then please reopen this issue and we'll continue to test. If you get a different problem with DisplayMagician v2.4.0  please create a new issue so I can track them separately.

Thanks
Terry
